File: ui-GLCamerax/src/main/java/com/melody/opengl/camerax/filters/gpuFilters/baseFilter/MagicBeautyFilter2.java
package com.melody.opengl.camerax.filters.gpuFilters.baseFilter;

import android.opengl.GLES20;

import com.melody.opengl.camerax.R;
import com.melody.opengl.camerax.filters.gpuFilters.utils.OpenGlUtils;

public class MagicBeautyFilter2 extends GPUImageFilter {
    private int mSingleStepOffsetLocation;
    private int mParamsLocation;
    private int mBrightnessLocation;
    private float mToneLevel;
    private float mBeautyLevel;
    private float mBrightLevel;

    public MagicBeautyFilter2(){
        super(NO_FILTER_VERTEX_SHADER ,
                OpenGlUtils.readShaderFromRawResource(R.raw.beauty2));
    }

    protected void onInit() {
        // 测试效果的参数值
        /*mToneLevel = 0.5F;
        mBeautyLevel = 1F;
        mBrightLevel = 0.2F;*/
        super.onInit();
        mParamsLocation = GLES20.glGetUniformLocation(getProgram(), "params");
        mBrightnessLocation = GLES20.glGetUniformLocation(getProgram(), "brightness");
        mSingleStepOffsetLocation = GLES20.glGetUniformLocation(getProgram(), "singleStepOffset");
        setParams(mBeautyLevel, mToneLevel);
        setBrightLevel(mBrightLevel);
    }

    private void setTexelSize(final float w, final float h) {
        setFloatVec2(mSingleStepOffsetLocation, new float[] {2.0f / w, 2.0f / h});
    }

    @Override
    public void onInputSizeChanged(final int width, final int height) {
        super.onInputSizeChanged(width, height);
        setTexelSize(width, height);
    }

    public float getBeautyLevel(){
        return this.mBeautyLevel;
    }

    public boolean getBeautyFeatureEnable() {
        return mToneLevel > 0F || mBeautyLevel > 0F;
    }

    /**
     * 磨皮
     */
    public void setBeautyLevel(float beautyLevel) {
        this.mBeautyLevel = beautyLevel;
        setParams(beautyLevel, mToneLevel);
    }

    /**
     * 美白
     */
    public void setBrightLevel(float brightLevel) {
        this.mBrightLevel = brightLevel;
        setFloat(mBrightnessLocation, 0.6f * (-0.5f + brightLevel));
    }

    /**
     * 红润
     */
    public void setToneLevel(float toneLevel) {
        this.mToneLevel = toneLevel;
        setParams(mBeautyLevel, toneLevel);
    }

    public void setAllBeautyParams(float beauty,float bright,float tone) {
        setBeautyLevel(beauty);
        setBrightLevel(bright);
        setToneLevel(tone);
    }

    private void setParams(float beauty, float tone) {
        float[] vector = new float[4];
        vector[0] = 1.0f - 0.6f * beauty;
        vector[1] = 1.0f - 0.3f * beauty;
        vector[2] = 0.1f + 0.3f * tone;
        vector[3] = 0.1f + 0.3f * tone;
        setFloatVec4(mParamsLocation, vector);
    }
}

File: ui-GLCamerax/src/main/java/com/melody/opengl/camerax/filters/gpuFilters/baseFilter/MagicBeautyFilter2.kt
package com.melody.opengl.camerax.filters.gpuFilters.baseFilter

import com.melody.opengl.camerax.filters.gpuFilters.utils.OpenGlUtils.readShaderFromRawResource
import com.melody.opengl.camerax.R
import android.opengl.GLES20

class MagicBeautyFilter2 : GPUImageFilter(
    NO_FILTER_VERTEX_SHADER,
    readShaderFromRawResource(R.raw.beauty2)
) {
    private var mSingleStepOffsetLocation = 0
    private var mParamsLocation = 0
    private var mBrightnessLocation = 0
    private var mToneLevel = 0f
    private var mBeautyLevel = 0f
    private var mBrightLevel = 0f
    override fun onInit() {
        // 测试效果的参数值
        /*mToneLevel = 0.5F;
        mBeautyLevel = 1F;
        mBrightLevel = 0.2F;*/
        super.onInit()
        mParamsLocation = GLES20.glGetUniformLocation(program, "params")
        mBrightnessLocation = GLES20.glGetUniformLocation(program, "brightness")
        mSingleStepOffsetLocation = GLES20.glGetUniformLocation(program, "singleStepOffset")
        setParams(mBeautyLevel, mToneLevel)
        setBrightLevel(mBrightLevel)
    }

    private fun setTexelSize(w: Float, h: Float) {
        setFloatVec2(mSingleStepOffsetLocation, floatArrayOf(2.0f / w, 2.0f / h))
    }

    override fun onInputSizeChanged(width: Int, height: Int) {
        super.onInputSizeChanged(width, height)
        setTexelSize(width.toFloat(), height.toFloat())
    }

    /**
     * 磨皮
     */
    var beautyLevel: Float
        get() = mBeautyLevel
        set(beautyLevel) {
            mBeautyLevel = beautyLevel
            setParams(beautyLevel, mToneLevel)
        }
    val beautyFeatureEnable: Boolean
        get() = mToneLevel > 0f || mBeautyLevel > 0f

    /**
     * 美白
     */
    fun setBrightLevel(brightLevel: Float) {
        mBrightLevel = brightLevel
        setFloat(mBrightnessLocation, 0.6f * (-0.5f + brightLevel))
    }

    /**
     * 红润
     */
    fun setToneLevel(toneLevel: Float) {
        mToneLevel = toneLevel
        setParams(mBeautyLevel, toneLevel)
    }

    fun setAllBeautyParams(beauty: Float, bright: Float, tone: Float) {
        beautyLevel = beauty
        setBrightLevel(bright)
        setToneLevel(tone)
    }

    private fun setParams(beauty: Float, tone: Float) {
        val vector = FloatArray(4)
        vector[0] = 1.0f - 0.6f * beauty
        vector[1] = 1.0f - 0.3f * beauty
        vector[2] = 0.1f + 0.3f * tone
        vector[3] = 0.1f + 0.3f * tone
        setFloatVec4(mParamsLocation, vector)
    }
}